### PR TITLE
[color-string] Use more appropriate defs for nullish inputs

### PR DESCRIPTION
**Pull request by ccjmne**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

### CODE REVIEW

**Comment by ccjmne** [comment]
The previous definition would allow passing *no* value to the `get.hsl`, `get.rwb` and `get.rgb` functions, which, while effectively possible, was not sensible (as hinted by @BendingBender).

Merely accepting `string | null` for the three declinations of the `get` function simply works and does error out at a `get.rgb()` call without any argument.

---

Also, I really shot myself in the foot before: the tests didn't reflect the actual use-case and the definitions wouldn't even work.

Assuming the following definitions:
```typescript
function rgb(colorString: string): Color | null;
function rgb(empty: null | undefined): null;
```

This call would not match any overload... 🤦‍♀️
```typescript
function testNullableInput(input: string | null) {
    color = colorString.get.rgb(input); // TS2769: No overload matches this call 
}
```

**Comment by ccjmne** [comment]
In the end, all I did boils down to:

```diff
-function hsl(colorString: string): Color | null;
+function hsl(colorString: string | null): Color | null;

-function hwb(colorString: string): Color | null;
+function hwb(colorString: string | null): Color | null;

-function rgb(colorString: string): Color | null;
+function rgb(colorString: string | null): Color | null;
```

Sorry for the wonky implementation the first time around! I'll be more careful in the future :)

File: types/color-string/index.d.ts [hunk 1]
```diff
@@ -15,12 +15,9 @@ export interface ColorDescriptor {
...
 export namespace get {
-    function hsl(colorString: string): Color | null;
-    function hsl(empty: null | undefined): null;
-    function hwb(colorString: string): Color | null;
-    function hwb(empty: null | undefined): null;
-    function rgb(colorString: string): Color | null;
-    function rgb(empty: null | undefined): null;
+    function hsl(colorString: string | null): Color | null;
+    function hwb(colorString: string | null): Color | null;
+    function rgb(colorString: string | null): Color | null;
```
Comment (BendingBender) [inline types/color-string/index.d.ts:20]: I'd implement it differently. This definition allows to pass in non-nullable types and get back non-nullable results while allowing passing in nullable types.
```suggestion
    function hsl<TColorStr extends string | null>(colorString: TColorStr): TColorStr extends string ? Color : null;
    function hwb<TColorStr extends string | null>(colorString: TColorStr): TColorStr extends string ? Color : null;
    function rgb<TColorStr extends string | null>(colorString: TColorStr): TColorStr extends string ? Color : null;
```

File: types/color-string/color-string-tests.ts [hunk 2]
```diff
@@ -21,10 +21,12 @@ color = colorString.get.hwb('hwb(60, 3%, 60%, 0.6)');
...
-let nothing: null;
-nothing = colorString.get.rgb(null);
-nothing = colorString.get.hsl(null);
-nothing = colorString.get.hwb(null);
+function testNullableInput(input: string | null) {
+    color = colorString.get.rgb(input);
+}
+
+// $ExpectError
+color = colorString.get.rgb();
```
Comment (BendingBender) [inline types/color-string/color-string-tests.ts:29]: The following tests should test the new functionality:
```suggestion
colorString.get.rgb(null); // $ExpectType null
colorString.get.rgb('hwb(60, 3%, 60%)'); // $ExpectType Color
colorString.get.rgb(null as string | null); // $ExpectType Color | null
colorString.get.rgb(); // $ExpectError

colorString.get.hwb(null); // $ExpectType null
colorString.get.hwb('hwb(60, 3%, 60%)'); // $ExpectType Color
colorString.get.hwb(null as string | null); // $ExpectType Color | null
colorString.get.hwb(); // $ExpectError

colorString.get.hsl(null); // $ExpectType null
colorString.get.hsl('hwb(60, 3%, 60%)'); // $ExpectType Color
colorString.get.hsl(null as string | null); // $ExpectType Color | null
colorString.get.hsl(); // $ExpectError
```

**Review by BendingBender — CHANGES_REQUESTED** [review]

**Comment by ccjmne** [comment]
It never occurred to me that conditional return types were possible! Thank you very much, it's very neat :)

I didn't write exhaustive tests initially so as to not deviate much from what's essentially just the "Usage" documentation from the Javascript implementation (https://github.com/Qix-/color-string#readme), but I'm all for it.

Thank you again for this!

**Review by BendingBender — APPROVED** [review]

**Comment by ccjmne** [comment]
Ready to merge

**Pull request merged**